From e278e859dbdfed1b3145f3af7558eb2956dd6061 Mon Sep 17 00:00:00 2001
Message-ID: <e278e859dbdfed1b3145f3af7558eb2956dd6061.1772540698.git.gerald.elder-vass@citrix.com>
From: Gerald Elder-Vass <gerald.elder-vass@citrix.com>
Date: Tue, 3 Mar 2026 12:18:43 +0000
Subject: [PATCH] CA-424241: Handle strings for usb hostbus property

USB devices expect an integer for the hostbus property.
Before bd5629d (v9.2.0), an upstream bug allowed a string to be
interpreted as an integer. The toolstack currently treats hostbus as a
string and our version of the OCaml QMP library also uses a string.

The short term fix is to continue supporting strings for the hostbus
property by replacing the string with the integer equivalent before
using the property.

The toolstack can be updated to use the updated library and integers
while QEMU handles both cases (and then the string support can be
removed).

Signed-off-by: Gerald Elder-Vass <gerald.elder-vass@citrix.com>
---
 system/qdev-monitor.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/system/qdev-monitor.c b/system/qdev-monitor.c
index 2ac92d0a076c..3bcbe4337feb 100644
--- a/system/qdev-monitor.c
+++ b/system/qdev-monitor.c
@@ -631,6 +631,7 @@ DeviceState *qdev_device_add_from_qdict(const QDict *opts,
     DeviceState *dev;
     BusState *bus = NULL;
     QDict *properties;
+    int hostbus;
 
     driver = qdict_get_try_str(opts, "driver");
     if (!driver) {
@@ -704,6 +705,13 @@ DeviceState *qdev_device_add_from_qdict(const QDict *opts,
     qdict_del(properties, "bus");
     qdict_del(properties, "id");
 
+    /* convert hostbus from str to int if necessary */
+    if (qdict_get_try_str(properties, "hostbus") != NULL) {
+        hostbus = atoi(qdict_get_str(properties, "hostbus"));
+        qdict_del(properties, "hostbus");
+        qdict_put_int(properties, "hostbus", hostbus);
+    }
+
     object_set_properties_from_keyval(&dev->parent_obj, properties, from_json,
                                       errp);
     qobject_unref(properties);
-- 
2.52.0

